### PR TITLE
Add support for cross-realm Kerberos authentication

### PR DESCRIPTION
**Pull request by arhimondr**

### CODE REVIEW

**Comment by arhimondr** [comment]
CC: @phd1994 @findepi

**Comment by arhimondr** [comment]
@phd1994 merged! Sorry for the delay, and thanks for your contribution!

File: presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosHadoopAuthentication.java [hunk 1]
```diff
@@ -25,24 +26,36 @@
...
-    static {
+    private final KerberosAuthentication kerberosAuthentication;
+
+    public static KerberosHadoopAuthentication createKerberosHadoopAuthentication(KerberosAuthentication kerberosAuthentication, HdfsConfiguration hdfsConfiguration)
+    {
+        KerberosHadoopAuthentication kerberosHadoopAuthentication = new KerberosHadoopAuthentication(kerberosAuthentication);
+
+        Configuration configuration = requireNonNull(hdfsConfiguration, "hdfsConfiguration is null").getConfiguration();
+        initializeConfigs(configuration);
+        return kerberosHadoopAuthentication;
```
Comment (sopel39) [inline presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosHadoopAuthentication.java]: How `kerberosHadoopAuthentication` is bound with `hdfsConfiguration`? They are not used in the same context within this method. Is this via some static variables?

Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosHadoopAuthentication.java]: Unfortunately yes.

File: presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosHadoopAuthentication.java [hunk 2]
```diff
@@ -25,24 +26,36 @@
...
         // It is further used in many places in DfsClient
-        Configuration configuration = new Configuration(false);
         configuration.set("hadoop.security.authentication", "kerberos");
-        UserGroupInformation.setConfiguration(configuration);
 
-        // KerberosName#rules static field must be initialized
-        // It is used in KerberosName#getShortName which is used in User constructor invoked by UserGroupInformation#getUGIFromSubject
-        KerberosName.setRules("DEFAULT");
-    }
+        String authToLocalRules = configuration.get("hadoop.security.auth_to_local");
```
Comment (sopel39) [inline presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosHadoopAuthentication.java]: why not use `org.apache.hadoop.security.HadoopKerberosName#setConfiguration` instead?

Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosHadoopAuthentication.java]: When i was implementing this, i wanted to get to the bottom of that nasty code, and leave it as obvious as possible. That's why i didn't use `HadoopKerberosName#setConfiguration`. Please don't change it.

Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosHadoopAuthentication.java]: This method is actually being called from the inside of the `UserGroupInformation.setConfiguration`. By setting the rules to `DEFAULT` explicitly i wanted to explicitly depict the statistic components, that are being set. You can just remove the `KerberosName.setRules("DEFAULT")` all together.

**Comment by phd1994** [comment]
@arhimondr Thanks for merging. Appreciate your reviews and comments!

**Pull request closed without merging**